tests/fork-change-slot.c

```diff
@@ -0,0 +1,347 @@
+}
+
+static void error_queue(const char *name, int pid)
+{
+    if (ERR_peek_last_error()) {
+        fprintf(stderr, "pid %d: %s generated errors:\n", pid, name);
+        ERR_print_errors_fp(stderr);
+    }
+}
+
+static X509 *read_cert(char *file, int pid)
+{
+    BIO *bio = NULL;
+    X509 *cert = NULL; 
+
+    bio = BIO_new(BIO_s_file());
+
+    if (bio == NULL) {
+        error_queue("BIO_new", pid);
+        return NULL;
+    }
+
+    if (BIO_read_filename(bio, file) < 0) {
+        error_queue("BIO_read_filename", pid);
+        BIO_free(bio);
+        return NULL;
+    }
+
+    cert = PEM_read_bio_X509(bio, NULL, NULL, NULL);
+
+    if (cert == NULL) {
+        error_queue("PEM_read_bio_X509", pid);
+    }
+
+    BIO_free(bio);
+    return cert;
+}
+
+static void usage(char *argv[])
+{
+    printf("usage: %s (Key PKCS#11 URL) (Certificate file path) (# of spawned processes) [opt: PKCS#11 module path]\n",
+            argv[0]);
+}
+
+int main(int argc, char *argv[])
+{
+    const EVP_MD *digest_algo = NULL;
+    EVP_PKEY *privkey = NULL, *pubkey = NULL;
+    EVP_MD_CTX *md_ctx = NULL;
+    ENGINE *engine = NULL;
+    X509 *cert = NULL;
+    unsigned char *random = NULL, *signature = NULL;
+
+    int rc = 0, fd, num_processes = 0;
+    unsigned int siglen;
+    pid_t pid;
+
+    int rv = 1;
+
+    /* Check arguments */
+    if (argc < 4) {
+        fprintf(stderr, "Missing required arguments\n");
+        usage(argv);
+        goto failed;
+    }
+
+    if (argc > 5) {
+        fprintf(stderr, "Too many arguments\n");
+        usage(argv);
+        goto failed;
+    }
+
+    /* Check PKCS#11 URL */
+    if (strncmp(argv[1], "pkcs11:", 7)) {
+        fprintf(stderr, "fatal: invalid PKCS#11 URL\n");
+        usage(argv);
+        goto failed;
+    }
+
+    /* Get the number of processes to be spawned */
+    num_processes = atoi(argv[3]);
+    if ((num_processes <= 0) || (num_processes > 100)) {
+        fprintf(stderr, "fatal: invalid number of processes (use from 1 to 100)\n");
+        goto failed;
+    }
+
+    ENGINE_load_builtin_engines();
+
+    pid = getpid();
+
+    printf("pid %d is the parent\n", pid);
+
+    /* Read certificate file */
+    cert = read_cert(argv[2], pid);
+    if (cert == NULL) {
+        error_queue("read_cert", pid);
+        goto failed;
+    }
+
+    /* Get the public key for verification */
+    pubkey = X509_get_pubkey(cert);
+    if (pubkey == NULL) {
+        fprintf(stderr, "pid %d: Could not extract the public key\n", pid);
+        error_queue("X509_get_pubkey", pid);
+        goto failed;
+    }
+
+    /* Get structural reference */
+    engine = ENGINE_by_id("pkcs11");
+    if (engine == NULL) {
+        fprintf(stderr, "fatal: engine \"pkcs11\" not available\n");
+        error_queue("ENGINE_by_id", pid);
+        goto failed;
+    }
+
+    /* Initialize to get the engine functional reference */
+    if (ENGINE_init(engine)) {
+        //TODO Improve arguments parsing using getopt or argp
```

Comment:
I don't think it makes sense to have a todo in code which you send upstream. I would look like a todo for him. About the point itself, I'm not sure what would a full blown parser buy on a test suite tool.

Reply:
About the point, the main reason was to use automatic generation of options like "-h" for help. But, since this code is for testing purposes, and it has a specific goal, it seems unnecessary. I'll remove all the TODO's in future codes that I am willing to submit to the upstream.